Support Top in slotted runtime

henriknyman opened the pull request:
This also fixes a bug that `SortSlottedPipe` did not handle long slots correctly.

henriknyman committed:
Refactor SortPipe

henriknyman committed:
Add Top to slotted runtime

henriknyman committed:
Remove old implementation from TopSlottedPipe

henriknyman committed:
Port Top unit tests to slotted runtime

henriknyman committed:
Refine unit tests

henriknyman commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/PrimitiveExecutionContext.scala:
```diff
@@ -80,7 +80,7 @@ case class PrimitiveExecutionContext(pipeline: PipelineInformation) extends Exec
     // This method implementation is for debug usage only (the debugger will invoke it when stepping).
     // Please do not use in production code.
-    (longs.map(i => ("LongSlot", ValueUtils.of(i))) ++ refs.map(i => ("RefSlot", ValueUtils.of(i)))).iterator
+    (longs.map(i => ("LongSlot", ValueUtils.of(i))) ++ refs.map(i => ("RefSlot", i))).iterator
```
I think this change can be reverted

henriknyman replied:
No, it actually needs to be there to avoid exceptions during debugging

henriknyman committed:
Remove passing tests from blacklist

Minor cleanup

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/SlottedPipeBuilder.scala:
```diff
@@ -226,6 +226,20 @@ class SlottedPipeBuilder(fallback: PipeBuilder,
+      case Top(_, sortItems, limit) =>
+        TopNSlottedPipe(source, sortItems.map(translateColumnOrder(pipeline, _)).toList, convertExpressions(limit))(id = id)
+
+      case Limit(_, count, IncludeTies) =>
```
Why is selection of Top vs Sort+Limit not a planner decision here too?

henriknyman replied:
I don't know

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/SlottedPipeBuilder.scala:
```diff
@@ -226,6 +226,20 @@ class SlottedPipeBuilder(fallback: PipeBuilder,
+
+      case Limit(_, count, IncludeTies) =>
+        (source, count) match {
+          case (SortSlottedPipe(inner, sortDescription, _), SignedDecimalIntegerLiteral("1")) =>
```
Is WithTies only valid for Limit=1, or do we just not want to have specializations for TopNWithTies?

henriknyman replied:
It is a special case only used internally by shortest path planning

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/pipes/TopSlottedPipe.scala:
```diff
@@ -0,0 +1,157 @@
+    }
+  }
+
+  def binarySearch(array: Array[ExecutionContext], comparator: Comparator[ExecutionContext])(key: ExecutionContext) = {
```
Now that we use `DefaultComparatorTopTable` this is never used

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/pipes/SortSlottedPipe.scala:
```diff
@@ -30,18 +30,20 @@ case class SortSlottedPipe(source: Pipe, orderBy: Seq[ColumnOrder], pipelineInfo
 
-private class ExecutionContextOrdering(order: ColumnOrder)(implicit qtx: QueryState) extends scala.Ordering[ExecutionContext] {
+private class ExecutionContextOrdering(order: ColumnOrder) extends scala.Ordering[ExecutionContext] {
+  val columnSlot: Int = order.slot
```
A bit OCD, but for consistency with TopSlottedPipe this would be private

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/main/java/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/DefaultComparatorTopTable.java:
```diff
@@ -0,0 +1,128 @@
+    private final int totalCount;
+    private int count = -1;
+    private PriorityQueue<T> heap;
+    private Object[] array; // TODO: Use Guava's MinMaxPriorityQueue to avoid having this array
```
Removing the need for this array would indeed be nice

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/pipes/TopSlottedPipe.scala:
```diff
@@ -0,0 +1,157 @@
+
+  @inline
+  private def init(first: ExecutionContext) = {
+    val builder = Vector.newBuilder[ExecutionContext]
```
Have you tested if something like `ListBuffer` or `ArrayBuilder` would be faster?

henriknyman replied:
No. I don't think we need to optimize for this case right now..

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/pipes/TopSlottedPipe.scala:
```diff
@@ -0,0 +1,157 @@
+          val comparison = comparator.compare(ctx, best)
+          if (comparison < 0) { // Found a new best
+            best = ctx
+            matchingRows = init(ctx)
```
Presumably this could happen very often. 
Can we do `matchingRows.clear()` instead?
(although `VectorBuilder` does very similar things in constructor as it does in `clear()` at least we'd create fewer `VectorBuilder` instances)

henriknyman replied:
Sure

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/test/java/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/DefaultComparatorTopTableTest.java:
```diff
@@ -0,0 +1,139 @@
+    public final ExpectedException exception = ExpectedException.none();
+
+    @Test
+    public void shouldHandleAddingMoreValuesThenCapacity()
```
`...ValuesThenCapacity...` -> `...ValuesThanCapacity...`

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/test/java/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/DefaultComparatorTopTableTest.java:
```diff
@@ -0,0 +1,139 @@
+/*
```
Looks like this class (and perhaps others) hasn't been auto formatted

henriknyman committed:
Fix slotted sort and top pipes to handle long slots

Slotted Sort and Top pipes need to track slot type in the ColumnOrder

henriknyman committed:
Fix spec suite test to run with specific feature

Spec suite test blacklist diff verification needs to be disabled if
we are running a specific feature and/or scenario.

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/pipes/Top1WithTiesSlottedPipeTest.scala:
```diff
@@ -0,0 +1,123 @@
+    sortPipe.createResults(QueryStateHelper.emptyWithValueSerialization).toList should beEquivalentTo(List(Map("x" -> "A")))
+  }
+
+  test("three ties for the first place are all returned") {
```
"three ties for the first..." --> "two ties for the first"

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/pipes/Top1WithTiesSlottedPipeTest.scala:
```diff
@@ -0,0 +1,123 @@
+      Map("x" -> null, "y" -> 2)))
+  }
+
+  test("null should not be returned if other values are present") {
```
why is `null` treated as some special case?
isn't this entirely up to what the CIP specifies: https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2016-06-14-Define-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc#orderability-equivalence

henriknyman replied:
This test precedes both the new value model and that CIP

alexaverbuch commented on enterprise/cypher/slotted-runtime/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/pipes/TopSlottedPipeTest.scala:
```diff
@@ -0,0 +1,254 @@
+
+class TopSlottedPipeTest extends CypherFunSuite {
+
+  private sealed trait _ColumnOrder
```
since when do we use `_` prefix for private?

henriknyman committed:
Fix some review comments

henriknyman committed:
Blacklist test in compatibility mode

henriknyman committed:
Simplify unit tests

craigtaverner commented on enterprise/cypher/slotted-runtime/src/main/java/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/DefaultComparatorTopTable.java:
```diff
@@ -0,0 +1,128 @@
+import java.util.PriorityQueue;
+
+/**
+ * The default implementation of a Top N table used by the generated code
```
'generated code'?

craigtaverner commented on enterprise/cypher/slotted-runtime/src/main/java/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/DefaultComparatorTopTable.java:
```diff
@@ -0,0 +1,128 @@
+ * It accepts tuples as boxed objects that implements Comparable
+ *
+ * Implements the following interface:
+ * (since the code is generated it does not actually need to declare it with implements)
```
This comment seems to have been copied from classes in the compiled runtime. Should be reworded or deleted.

craigtaverner commented on enterprise/cypher/slotted-runtime/src/main/java/org/neo4j/cypher/internal/compatibility/v3_3/runtime/slotted/DefaultComparatorTopTable.java:
```diff
@@ -0,0 +1,128 @@
+    {
+        if ( count == -1 )
+        {
+            // This should never happen in generated code but is here to simplify debugging if used incorrectly
```
Another compiled runtime comment?
